### PR TITLE
add thrift wrapper for meta and storage

### PR DESCRIPTION
**Pull request by jievince**
Add some thrift wrapper about reading  vertexes, edges and meta information.
test on a 3-node cluster:
192.168.8.16 is meta server ip, 45500 is meta listening port, nba is the space name to read.
run example `python3 GetFullGraph.py 192.168.8.16 45500 nba`
result:
```shell
scaning space nba
ok
tagName:  player
vid:  109
{'name': 'Tiago Splitter', 'age': 34}
vid:  119
{'name': 'Kevin Durant', 'age': 30}
vid:  129
{'name': 'Dwyane Wade', 'age': 37}
vid:  139
{'name': 'Marc Gasol', 'age': 34}
vid:  149
{'name': 'Ben Simmons', 'age': 22}
tagName:  team
vid:  209
{'name': 'Timberwolves'}
vid:  219
{'name': '76ers'}
vid:  229
{'name': 'Heat'}
tagName:  player
vid:  108
{'name': 'Boris Diaw', 'age': 36}
vid:  118
{'name': 'Russell Westbrook', 'age': 30}
vid:  128
...
```

### CODE REVIEW

File: nebula/ngMeta/MetaClient.py [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+            return None
+        getEdgeReq = GetEdgeReq(spaceId, edgeName, -1)
+        getEdgeResp = self.metaClient.getEdge(getEdgeReq)
+        if getEdgeResp.code == ErrorCode.SUCCEEDED:
+            return getEdgeResp.Schema
+        else:
+            print('get edge error, error code: ', getEdgeResp.code)
+            return None
+
+    def getEdgeSchema(self, spaceName, edgeName, version=-1):
```
Comment (laura-ding) [inline nebula/ngMeta/MetaClient.py:273]: `getEdgeSchema` can call `getEdge` directly.
why need two interfaces?

File: nebula/ngStorage/StorageClient.py [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+    if self._hasnext:
+      result = self._thenext
+    else:
+      result = next(self.it)
+    self._hasnext = None
+    return result
+
+  def hasNext(self):
+    if self._hasnext is None:
+      try: self._thenext = next(self.it)
```
Comment (laura-ding) [inline nebula/ngStorage/StorageClient.py:35]: Add a line break

File: nebula/ngMeta/MetaClient.py [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+        host = address[0]
+        port = address[1]
+        tTransport = TSocket.TSocket(host, port)
+        tTransport.setTimeout(self.timeout)
+        tTransport = TTransport.TBufferedTransport(tTransport)
+        tProtocol = TBinaryProtocol.TBinaryProtocol(tTransport)
+        tTransport.open()
+        self.metaClient = Client(tProtocol)
+
+        for spaceIdName in self.listSpaces():
```
Comment (laura-ding) [inline nebula/ngMeta/MetaClient.py:64]: the schema just get when connect, if the user modifies, the cache not update

Comment (jievince) [reply nebula/ngMeta/MetaClient.py:93]: add a RepeatTimer to update schemas every 2s

File: nebula/ngMeta/MetaClient.py (hunk 1 above)
Comment (jievince) [reply nebula/ngMeta/MetaClient.py:296]: These 2 interfaces return a different result, I think these 2 interfaces are both useful for users

**Comment by jievince** [comment]
test_StorageClient.py failed. 
docker-compose maps the metad listening port to a random port, but the test requires a fixed metad port to connect.

**Comment by jievince** [comment]
test still failed.
https://github.com/jievince/nebula-python/runs/975833289?check_suite_focus=true

File: examples/StorageClientExample.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    edgeReturnCols = {}
+    for edgeItem in edgeItems:
+        edgeName = edgeItem.edge_name
+        edgeReturnCols[edgeName] = metaClient.getEdgeSchema(space, edgeName).keys()
+
+    return vertexReturnCols, edgeReturnCols
+
+
+if __name__ == '__main__':
+    metaClient = MetaClient([(sys.argv[1], sys.argv[2])])
```
Comment (laura-ding) [inline examples/StorageClientExample.py:101]: add `try except`

File: nebula/ngMeta/MetaClient.py [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+        spaceId = self.getSpaceIdFromCache(spaceName)
+        print('spaceId: ', spaceId)
+        if spaceId == -1:
+            return None
+        getTagReq = GetTagReq(spaceId, tagName, version)
+        getTagResp = self.client.getTag(getTagReq)
+        result = {}
+        for columnDef in getTagResp.schema.columns:
+            result[columnDef.name] = columnDef.type.type
+        print('result: ', result)
```
Comment (laura-ding) [inline nebula/ngMeta/MetaClient.py:248]: delete the debug log

File: nebula/ngMeta/MetaClient.py [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+        getTagResp = self.client.getTag(getTagReq)
+
+        if getTagResp.code == ErrorCode.SUCCEEDED:
+            return getTagResp.schema
+        else:
+            return None
+
+    def getTagSchema(self, spaceName, tagName, version=-1):
+        spaceId = self.getSpaceIdFromCache(spaceName)
+        print('spaceId: ', spaceId)
```
Comment (laura-ding) [inline nebula/ngMeta/MetaClient.py:240]: delete the debug log

File: nebula/ngData/data.py [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+#
+# This source code is licensed under Apache 2.0 License,
+# attached with Common Clause Condition 1.0, found in the LICENSES directory.
+
+from enum import Enum
+import struct
+import six
+
+class PropertyDef:
+    PropertyType = Enum('PropertyType', ('UNKNOWN', 'BOOL', 'INT', 'VID', 'FLOAT', 'DOUBLE', \
```
Comment (laura-ding) [inline nebula/ngData/data.py:13]: missing `TIMESTAMP` type

File: nebula/ngData/data.py [hunk 8]
```diff
@@ -0,0 +1,184 @@
+# --coding:utf-8--
+#
+# Copyright (c) 2019 vesoft inc. All rights reserved.
```
Comment (laura-ding) [inline nebula/ngData/data.py:3]: 2020

File: examples/StorageClientExample.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
+# --coding:utf-8--
+#
+# Copyright (c) 2019 vesoft inc. All rights reserved.
```
Comment (laura-ding) [inline examples/StorageClientExample.py:3]: 2020

File: nebula/ngMeta/MetaClient.py [hunk 10]
```diff
@@ -0,0 +1,307 @@
+# --coding:utf-8--
+#
+# Copyright (c) 2019 vesoft inc. All rights reserved.
```
Comment (laura-ding) [inline nebula/ngMeta/MetaClient.py:3]: 2020

File: nebula/ngStorage/StorageClient.py [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+        self.cursor = None
+        self.haveNext = True
+
+    def hasNext(self):
+        return self.haveNext
+
+    def next(self):
+        self.scanVertexRequest.cursor = self.cursor
+        scanVertexResponse = self.client.scanVertex(self.scanVertexRequest)
+        assert(scanVertexResponse is not None), 'scanVertexReponse is none'
```
Comment (laura-ding) [inline nebula/ngStorage/StorageClient.py:98]: why assert?

File: nebula/ngData/data.py [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+        self.propertyType = propertyType
+        self.name = name
+
+
+class Property:
+    def __init__(self, propertyType, name, value):
+        self.propertyDef = PropertyDef(propertyType, name)
+        self.value = value
+    
+    def getType(self):
```
Comment (laura-ding) [inline nebula/ngData/data.py:26]: modify to `get_type`, the format of the function name

File: examples/StorageClientExample.py (hunk 4 above)
Comment (laura-ding) [reply examples/StorageClientExample.py:102]: MetaClient is possible to throw exceptions, such as thrift interface

File: nebula/ngData/data.py (hunk 12 above)
Comment (laura-ding) [reply nebula/ngData/data.py:26]: Yes, all functions.

File: nebula/ngData/data.py [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+        self._property_name_index = {}
+        self._property_types = []
+        self._offset = 0
+
+        prop_type_map = {0: PropertyDef.PropertyType.UNKNOWN, 1: PropertyDef.PropertyType.BOOL, 2: PropertyDef.PropertyType.INT,
+            3: PropertyDef.PropertyType.VID, 4: PropertyDef.PropertyType.FLOAT, 5: PropertyDef.PropertyType.DOUBLE,
+            6: PropertyDef.PropertyType.STRING, 21: PropertyDef.PropertyType.TIMESTAMP}
+        idx = 0
+        for column_def in schema.columns:
+            property_type = prop_type_map[column_def.type.type] # ColumnDef is in common/ttypes.py
```
Comment (laura-ding) [inline nebula/ngData/data.py:55]: I think you can use use  `SupportedType` directly. no need to define a map.

File: nebula/ngData/data.py [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+#
+# This source code is licensed under Apache 2.0 License,
+# attached with Common Clause Condition 1.0, found in the LICENSES directory.
+
+from enum import Enum
+import struct
+import six
+
+class PropertyDef:
+    PropertyType = Enum('PropertyType', ('UNKNOWN', 'BOOL', 'INT', 'VID', 'FLOAT', 'DOUBLE', \
```
Comment (laura-ding) [inline nebula/ngData/data.py:13]: why not use thrift define type directly?

File: tests/test_StorageClient.py [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+        time.sleep(5)
+        resp = client.execute('USE %s' % space_name)
+        assert resp.error_code == 0, resp.error_msg
+        time.sleep(5)
+        resp = client.execute('CREATE TAG player(name string, age int)')
+        assert resp.error_code == 0, resp.error_msg
+        resp = client.execute('CREATE EDGE follow(degree double)')
+        assert resp.error_code == 0, resp.error_msg
+        time.sleep(12)
+        resp = client.execute('INSERT VERTEX player(name, age) VALUES 1:(\'Bob\', 18)')
```
Comment (laura-ding) [inline tests/test_StorageClient.py:56]: You need to add tests to test all type which we have supported.

File: tests/test_StorageClient.py [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+    result = storage_client.scan_part_edge(space_name, 1, {'follow':['degree']}, True, 100, 0, sys.maxsize)
+    assert result is not None and result.next().edge_data is not None
+
+def test_scan_part_vertex():
+    result = storage_client.scan_part_vertex(space_name, 1, {'player':['name', 'age']}, True, 100, 0, sys.maxsize)
+    assert result is not None and result.next().vertex_data is not None
+
+def test_get_tag_schema():
+    result = meta_client.get_tag_schema(space_name, 'player')
+    expect = {'name':6, 'age':2}
```
Comment (laura-ding) [inline tests/test_StorageClient.py:85]: use the enum value name instead of it. It’s more intuitive.

File: nebula/ngMeta/MetaClient.py [hunk 17]
```diff
@@ -0,0 +1,304 @@
...
+            transport.setTimeout(self._timeout)
+            transport = TTransport.TBufferedTransport(transport)
+            protocol = TBinaryProtocol.TBinaryProtocol(transport)
+            transport.open()
+        except Exception as x:
+            print(x)
+            return -1
+        else:
+            self._client = Client(protocol)
+            self.update_schemas()
```
Comment (laura-ding) [inline nebula/ngMeta/MetaClient.py:89]: It also throws an exception

File: nebula/ngMeta/MetaClient.py [hunk 18]
```diff
@@ -0,0 +1,304 @@
...
+            # Loading edge schema's cache
+            edges = {}
+            edges_name = {}
+            for edge_item in self.get_edges(space_name):
+                edges[edge_item.edge_name] = edge_item
+                edges_name[edge_item.edge_type] = edge_item.edge_name
+            self._space_edge_items[space_name] = edges
+            self._edge_name_map[space_name] = edges_name
+        self.set_space_part_leader()
+        return 0
```
Comment (laura-ding) [inline nebula/ngMeta/MetaClient.py:119]: no need to return, it is void return function

File: tests/test_StorageClient.py [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+
+def test_get_edge_return_cols():
+    edge_item = meta_client.get_edge_item_from_cache(space_name, 'follow')
+    edge_type = edge_item.edge_type
+    entry_id = EntryId(edge_type=edge_type)
+    result = storage_client.get_edge_return_cols(space_name, {'follow':['degree', 'time']})
+    expect = {edge_type:[PropDef(PropOwner.EDGE, entry_id, 'degree'), PropDef(PropOwner.EDGE, entry_id, 'time')]}
+    assert result == expect
+
+def test_get_vertex_return_cols():
```
Comment (laura-ding) [inline tests/test_StorageClient.py:102]: need to test the result of data with all type

**Review by laura-ding — APPROVED** [review]
Excellent!

**Pull request merged**